src/HttpUpgrade.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 1996-2019 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_HTTP_UPGRADE_H
+#define SQUID_HTTP_UPGRADE_H
+
+#include "acl/forward.h"
+#include "base/StringView.h"
+#include "sbuf/forward.h"
+
+#include <map>
+
+/// a reference to a protocol name[/version] string; no 0-termination is assumed
+class ProtocolView
```

Comment:
This replicates the intended use of existing AnyP::ProtocolVersion class. Please simply extend that class with a ProtocolOther child to store the SBuf image of PROTOCOL_UNKNOWN protocols.

Reply:
I am thinking this demonstrates the issue I have with this feature. That Squid should:
 a) only be allowing Upgrade for protocols it has Client logic to handle properly, and
 b) access controls for admin to *forbid* Upgrade as needed, and
 c) continue to handle unknown protocols as unknown (not Upgrade'able)

The use of ProtocolVersion (with SBuf for PROTOCOL_UNKNOWN strings) aids this by only producing a non-PROTOCOL_UNKNOWN for protocols it is likely to be able to Upgrade.

NP: this ProtocolView detail is architectural. So I do not insist on this item being resolve to approve.

Reply:
> this ProtocolView detail is architectural. So I do not insist on this item being resolve to approve.

Thank you. I would very much prefer to avoid merging architectural problems, so I will take one more stab at understanding and addressing your concern. However, if that fails, I would suggest using a real-time discussion to resolve our differences. The latter is often recommended when facing long, protracted email/PR discussions.

----

> I am thinking this demonstrates the issue I have with this feature. That Squid should:

> a) only be allowing Upgrade for protocols it has Client logic to handle properly, and

AFAICT, we are in agreement here, and the above is already implemented: The admin told Squid that tunneling is the appropriate (for this Squid instance) handling of this protocol. Squid already knows how to tunnel, of course.

Now, I am guessing that you meant something else by "handle properly" -- you probably meant that there is some hard-coded logic in Squid code specific to that protocol. In that case, you are arguing that Squid should essentially have a hard-coded `if (ClientRequestsUpgradeToWebSocket()) tunnel()` code. Given the following facts, I do not think that it is a good idea to hard-code that logic (today):

1. WebSocket is not the only use case for supporting Upgrades. In some cases, there is no (open source) knowledge about the protocols (except that they can be safely tunneled).
2. WebSocket clients may use obscure/non-standard names to request an Upgrade. In most cases, the clients think they are talking to the origin server, not a proxy.
3. Default-to-tunnel would probably still be the wrong default for Squid behavior because it is such an easy abuse/bypass vector. If so, we would still need a configuration option to explicitly allow tunneling.

If we do hard-code ClientRequestsUpgradeToWebSocket(), then the only way Squid admins will be able to compensate for (or remain compatible with) the above facts is via squid.conf rules. Thus, ClientRequestsUpgradeToWebSocket() would have to be very similar to what this PR implements today as far as upgrade controls and logic are concerned. In other words, we would still be doing pretty much the same stuff this PR does, but _pretend_ that it is all WebSockets. I do not think that "pretend it is all WebSockets while accommodating all other real-world cases" is a good strategy here.


> b) access controls for admin to _forbid_ Upgrade as needed, and

Knowing the protocol does not imply it is a good idea to Upgrade to it by default. Today, Squid never forwards protocols that it is not explicitly configured to forward. For example, we do not start listening for FTP traffic on the defauilt FTP port just because we support FTP. Instead, we wait for the Squid admin to configure ftp_port. I think we should continue to do that on the Upgrade path. What are the advantages of changing the default to "allow all WebSocket traffic" here?

> c) continue to handle unknown protocols as unknown (not Upgrade'able)

Yes, same as in (a): We already do that if "unknown" is interpreted as "unknown to admin" and see above for why I think hard-coding that knowledge is not the best solution here.

 
> The use of ProtocolVersion (with SBuf for PROTOCOL_UNKNOWN strings) aids this by only producing a non-PROTOCOL_UNKNOWN for protocols it is likely to be able to Upgrade.

I am not sure I understand this point, but I think that _if_ we use your model and hard-code, say, WebSocket support into Squid, then we should use the _existing_ AnyP::ProtocolVersion class for WebSocket traffic -- it just becomes one of the protocols that Squid recognizes and supports natively. Furthermore, if we always hard-code Upgrade protocols (impractical IMO, but let's assume that we have done that for the sake of the argument), then there will be no need for the ProtocolVersion class at all! That new class is only needed if we want to handle Upgrade protocols differently from protocols with hard-coded client support (and, as discussed above, I think we do want that!).